src/dodal/devices/fast_grid_scan.py

```diff
@@ -266,25 +265,81 @@ async def complete(self):
             raise
 
     @abstractmethod
-    def _create_position_counter(self, prefix: str) -> SignalRW[int]:
-        pass
+    def _create_expected_images_signal(self) -> SignalR[int]: ...
 
+    @abstractmethod
+    def _create_position_counter(self, prefix: str) -> SignalRW[int]: ...
 
-class ZebraFastGridScan(FastGridScanCommon[ZebraGridScanParams]):
-    """Device for standard Zebra FGS. In this scan, the goniometer's velocity profile follows a parabolic shape between X steps,
-    with the slowest points occuring at each X step.
+    # This can be created within init rather than as a separate method after https://github.com/DiamondLightSource/mx-bluesky/issues/1203
+    @abstractmethod
+    def _create_scan_invalid_signal(self, prefix: str) -> SignalR[float]: ...
+
+    # This can be created within init rather than as a separate method after https://github.com/DiamondLightSource/mx-bluesky/issues/1203
+    @abstractmethod
+    def _create_motion_program(self, smargon_prefix: str) -> MotionProgram: ...
+
+
+class FastGridScanThreeD(FastGridScanCommon[ParamType]):
+    """Device for standard 3D FGS.
+
+    After completeing the first grid, if Z steps isn't 0, the goniometer will
+    rotate in the omega direction such that it moves from the X-Y, to the X-Z plane then
+    do a second grid scan. The detector is triggered after every x step.
+    See https://github.com/DiamondLightSource/hyperion/wiki/Coordinate-Systems for more.
+
+    Subclasses must implement _create_position_counter.
     """
 
     def __init__(self, prefix: str, name: str = "") -> None:
         full_prefix = prefix + "FGS:"
-        # Time taken to travel between X steps
-        self.dwell_time_ms = epics_signal_rw_rbv(float, f"{full_prefix}DWELL_TIME")
-
+        # Number of vertical steps during the second grid scan, after the rotation in omega
+        self.z_steps = epics_signal_rw_rbv(int, f"{prefix}Z_NUM_STEPS")
+        self.z_step_size = epics_signal_rw_rbv(float, f"{prefix}Z_STEP_SIZE")
+        self.z2_start = epics_signal_rw_rbv(float, f"{prefix}Z2_START")
+        self.y2_start = epics_signal_rw_rbv(float, f"{prefix}Y2_START")
         self.x_counter = epics_signal_r(int, f"{full_prefix}X_COUNTER")
         self.y_counter = epics_signal_r(int, f"{full_prefix}Y_COUNTER")
 
         super().__init__(full_prefix, prefix, name)
```

Comment:
Could: It's a bit odd to have the super call in the middle, maybe let's just put it at the start?